### ISSUE TITLE
ServerPublishStateModel#setServerPublishState sets a state that gets overwritten immediately

### ISSUE THREAD

**Issue by adietish** [redhat-developer/rsp-server#311]
```ServerPublishStateModel#setServerPublishState(int state, boolean fire)``` stores the publish state in an instance variable. If ```fire``` is ```true``` it'll then call ```fireState()```. 
```
	@Override
	public void setServerPublishState(int state, boolean fire) {
		if( state != this.publishState) {
			this.publishState = state;
			if( fire ) 
				fireState();
		}
	}
```
```fireState()``` re-evaluates the publishState according to the states of the deployables in ```#updateServerPublishStateFromDeployments```. 
```
	private void fireState() {
		updateServerPublishStateFromDeployments();
```
```
	private void updateServerPublishStateFromDeployments() {
		...
		setServerPublishState(newState, false);
```
```#updateServerPublishStateFromDeployments``` then overrides the value of ```state```that was given to ```#setServerPublishState(int state, boolean fire)```. 
It thus looks to me as if the ```state``` parameter in  ```#setServerPublishState(int state, boolean fire)``` doesn't make a lot of sense since it's overridden anyhow. Missing something?

**Comment by adietish** [comment]
quoted from @robstryker in https://github.com/redhat-developer/rsp-server/pull/306#issuecomment-488109191:

So... part of the issue here is that if a deployment goes out of sync, we want the server to notice that and mark the server state as out of sync. 

However, you are right:  If someone sets the server publish state to "incremental publish required", and then sets all modules to "synchronized",  the server publish state will be overridden, since all deployments are synchronized, there's no real reason why an incremental publish would be required. 

On the one hand, we should try to avoid overwriting the choices of the server adapter. On the other hand, we do want the server publish state to be kept accurate whenever possible. 

initialize, addDeployable, removeDeployable, fileChanged, and setServerPublishState all call this 'fire', which attempts to overwrite (or rather make more accurate) the server publish state. 

This is something worth fixing only if we can come up with a specific workflow that's incorrect, from the perspective of a server adapter writer. If for example you can come up with a series of calls that are reasonable for a server adapter to make somehow, that ends up with an intended state being overwritten, then we can consider it. 

I myself struggle to come up with such a workflow. I even try to be pretty open with my idea of what a server adapter might want to do or why, but, even this I find hard to find a flaw currently. 

If a new deployment is added (and not published), or removed (and not published), it seems obvious to me that the server itself would now require a full publish. If a file is changed, it seems very reasonable to update the server state from synchronized to whatever the maximum publish required by the modules are now.  

The only specific workflow this seems to fail is if the module set implies a one publish type (full, incremental, none), and the server adapter specifically tries to declare itself to require a different one. But I can't figure out why a server adapter writer might want to do that. 

If we really wanted to, we could move that call to updateServerPublishStateFromDeployments   out of the fireState, and back up into those 5 callers (with the exception of setServerPublishState).  

Example: 
  1) Remove updateServerPublishStateFromDeployments() from fireState()
  2) Add updateServerPublishStateFromDeployments() to addDeployable() before fireState() is called
  3) Add updateServerPublishStateFromDeployments() to removeDeployable() before fireState() is called
  4) Add updateServerPublishStateFromDeployments() to initialize() before fireState() is called
  5) Add updateServerPublishStateFromDeployments() to addDeployable() before fireState() is called
  6) do NOT add updateServerPublishStateFromDeployments() to setServerPublishState(etc) so as not to override an adapter's explicit decision?

**Comment by adietish** [comment]
@robstryker I see a much simpler solution:

expose publicly```#updateServerPublishStateFromDeployments()```, make ```#setServerPublishState``` private/protected and only use it internally.

I feel my suggestion confirmed in the following code in AbstractServerDelegate:

```
	protected void updatePublishStateFromModules() throws CoreException {
		// Clients override
		int maxState = ServerManagementAPIConstants.PUBLISH_STATE_NONE;
		List<DeployableState> states = getServerPublishModel().getDeployableStates();
		for( DeployableState s : states ) {
			maxState = (s.getPublishState() > maxState ? s.getPublishState() : maxState);
		}
		if( maxState > ServerManagementAPIConstants.PUBLISH_STATE_FULL)
			maxState = ServerManagementAPIConstants.PUBLISH_STATE_FULL;
		setServerPublishState(maxState);
	}
```

This code actually does the ~same as what's in ```ServerPublishStateModel#updateServerPublishStateFromDeployments```. It iterates over deployables in the publish model and sets the deployable set to the model, which then gets internally re-evaludated and overrides what the delegate was trying to set. If I'm right, then this code could be replaced by a simple call to the state model ```#updateServerPublishStateFromDeployments``` and lead to the same results.

**Comment by adietish** [comment]
This was implemented +/- as suggested in [commit cfbf1bfc036822d8b908688a420b214588d20f28](https://github.com/redhat-developer/rsp-server/commit/cfbf1bfc036822d8b908688a420b214588d20f28).
The code in ```AbstractServerDelegate#updatePublishStateFromModules``` was removed and replaced by a call to ```ServerPublishStateModel#updateServerPublishStateFromDeployments```.
```ServerPublishStateModel#fireState``` now does not call ```ServerPublishStateModel#updateServerPublishStateFromDeployments``` any more and thus  wont re-evaluate the server publish state (and override a given state) any more.
Closing as resolved.
